### DISCUSSION TITLE
Preprocessor snapshot is triggered twice

### DISCUSSION THREAD

**Discussion by phabos** [mobxjs/mobx-state-tree#2007] [Q&A] [answered]
Hello MST community,

I have question regarding the preProcessor methods. Here is a simple example below.
I notice that the preProcessor method is called twice, is it normal ? 
I was thinking that maybe because we load some payload in the model the preProcessor triggers only once ?
Thanks for the explanations !!

Here is a reproductible example
https://stackblitz.com/edit/js-yb4r7j?devToolsHeight=33&file=index.js

```js
import { types } from 'mobx-state-tree';
const ModelWithoutSnProc = types.model({
  someProp: types.string,
});

const Model = types.snapshotProcessor(ModelWithoutSnProc, {
  preProcessor(sn) {
    console.log('preprocessor', sn); // => two logs show up in the console
    return {
      someProp: sn.whateverApiProp,
    };
  },
});

const myModel = Model.create({ whateverApiProp: 'coucou' });
```

**Comment by coolsoftwaretyler** [comment]
It looks like this might be [expected/acceptable behavior](https://github.com/mobxjs/mobx-state-tree/issues/348)? It's an old thread, but it seems to suggest this is intended. However, I don't see how your example is modifying anything, so I can't quite explain the double-run.

I'm going to dig a little deeper here, but this isn't the first time folks have talked about this. Very interesting.

**Reply by coolsoftwaretyler** [reply, marked as answer]
Oh, I think I understand more after reading more carefully. 

The original post in that issue says: 

> First for typecheck second for instantiating. Both for create and applySnapshot. This is problem because transforming can be really slow.

And links to the relevant code. 

I would guess, since [MST typechecks in development but not production](https://mobx-state-tree.js.org/API/index#typecheck) that you could get around this by turning off the type checking for your model? Or that you might need to design your preProcess functions to be idempotent, as the responses in that thread conclude.

Does that help you at all?

**Reply by phabos** [reply]
Hey @coolsoftwaretyler Thanks a lot for you reply. 
ok i will check the type checking as you suggest, and see if i have this only in dev mode.
